sqlalchemy-stubs/sql/base.pyi

```diff
@@ -1,66 +1,82 @@
+    def argument_for(cls, dialect_name: str, argument_name: str, default: Any) -> None: ...
     @property
-    def dialect_kwargs(self): ...
+    def dialect_kwargs(self) -> _DialectArgView: ...
     @property
-    def kwargs(self): ...
+    def kwargs(self) -> _DialectArgView: ...
     @property
-    def dialect_options(self): ...
+    def dialect_options(self) -> PopulateDict: ...
 
 class Generative(object): ...
 
+_E = TypeVar('_E', bound=Executable)
+
 class Executable(Generative):
     supports_execution: bool = ...
-    def execution_options(self, **kw): ...
-    def execute(self, *multiparams, **params): ...
-    def scalar(self, *multiparams, **params): ...
+    def execution_options(self: _E, **kw: Any) -> _E: ...
+    def execute(self, *multiparams: Any, **params: Any) -> Optional[ResultProxy]: ...
+    def scalar(self, *multiparams: Any, **params: Any) -> Any: ...
     @property
-    def bind(self): ...
+    def bind(self) -> Optional[Union[Engine, Connection]]: ...
 
 class SchemaEventTarget(object): ...
 
 class SchemaVisitor(ClauseVisitor):
     __traverse_options__: Any = ...
 
 class ColumnCollection(util.OrderedProperties[Column]):
-    def __init__(self, *columns) -> None: ...
-    def replace(self, column): ...
-    def add(self, column): ...
-    def __delitem__(self, key): ...
-    def __setattr__(self, key: str, object: Column) -> None: ...
-    def __setitem__(self, key, value: Column): ...
-    def clear(self): ...
-    def remove(self, column): ...
-    def update(self, iter): ...
-    def extend(self, iter): ...
-    __hash__: Any = ...
-    def __eq__(self, other): ...
-    def __contains__(self, other): ...
-    def contains_column(self, col): ...
-    def as_immutable(self): ...
+    def __init__(self, *columns: Column) -> None: ...
+    def __str__(self) -> str: ...
+    def replace(self, column: Column) -> None: ...
+    def add(self, column: Column) -> None: ...
+    def __delitem__(self, key: str) -> NoReturn: ...
+    def __setattr__(self, key: str, object: Column) -> NoReturn: ...
+    def __setitem__(self, key: str, value: Column) -> None: ...
+    def clear(self) -> NoReturn: ...
+    def remove(self, column: Column) -> None: ...
+    def update(self, iter: Iterable[Column]) -> None: ...
+    def extend(self, iter: Iterable[Column]) -> None: ...
+    __hash__: None = ...  # type: ignore
+    def __eq__(self, other: Any) -> bool: ...
+    def __contains__(self, other: Any) -> bool: ...
+    def contains_column(self, col: Column) -> bool: ...
+    def as_immutable(self) -> ImmutableColumnCollection: ...
 
 class ImmutableColumnCollection(util.ImmutableProperties[Column], ColumnCollection):
-    def __init__(self, data, all_columns) -> None: ...
-    extend: Any = ...
-    remove: Any = ...
+    def __init__(self, data: Dict[str, Any], all_columns: List[Column]) -> None: ...
+    def extend(self, iter: Iterable[Column]) -> NoReturn: ...
+    def remove(self, column: Column) -> NoReturn: ...
```

Comment:
Are these explicitly documented or just present in code to give better error message? If the latter, then maybe just remove these two.

Reply:
These methods are set to `sqlalchemy.util._collections.ImmutableContainer._immutable` which throws. I imagine they can just be removed.